piet-common/src/cairo_back.rs

```diff
@@ -101,41 +101,84 @@ impl<'a> BitmapTarget<'a> {
+    /// and doesn't write anything.
+    pub fn copy_raw_pixels(
+        &mut self,
+        fmt: ImageFormat,
+        buf: &mut [u8],
+    ) -> Result<usize, piet::Error> {
         // TODO: convert other formats.
         if fmt != ImageFormat::RgbaPremul {
             return Err(piet::Error::NotSupported);
         }
-        std::mem::drop(self.cr);
         self.surface.flush();
         let stride = self.surface.get_stride() as usize;
         let width = self.surface.get_width() as usize;
         let height = self.surface.get_height() as usize;
-        let mut raw_data = vec![0; width * height * 4];
-        let buf = self
-            .surface
-            .get_data()
-            .map_err(Into::<Box<dyn std::error::Error>>::into)?;
-        for y in 0..height {
-            let src_off = y * stride;
-            let dst_off = y * width * 4;
-            for x in 0..width {
-                raw_data[dst_off + x * 4 + 0] = buf[src_off + x * 4 + 2];
-                raw_data[dst_off + x * 4 + 1] = buf[src_off + x * 4 + 1];
-                raw_data[dst_off + x * 4 + 2] = buf[src_off + x * 4 + 0];
-                raw_data[dst_off + x * 4 + 3] = buf[src_off + x * 4 + 3];
+        let size = width * height * 4;
+        if buf.len() < size {
+            return Err(piet::Error::InvalidInput);
+        }
+        unsafe {
+            // Cairo's rust wrapper has extra safety checks that we want to avoid: it won't let us
+            // get the data from an ImageSurface that's still referenced by a context. The C docs
+            // don't seem to think that's a problem, as long as we call flush (which we already
+            // did), and promise not to mutate anything.
+            // https://www.cairographics.org/manual/cairo-Image-Surfaces.html#cairo-image-surface-get-data
+            let data = {
+                let data_ptr = cairo_sys::cairo_image_surface_get_data(self.surface.to_raw_none());
+                if data_ptr.is_null() {
+                    let err = cairo::BorrowError::from(cairo::Status::SurfaceFinished);
+                    return Err((Box::new(err) as Box<dyn std::error::Error>).into());
+                }
+                std::slice::from_raw_parts(data_ptr, height.saturating_sub(1) * stride + width * 4)
+            };
+            for y in 0..height {
+                let src_off = y * stride;
+                let dst_off = y * width * 4;
+                for x in 0..width {
+                    // These unchecked indexes allow the autovectorizer to shine.
+                    // Note that dst_off maxes out at (height - 1) * width * 4, and so
+                    // dst_off + x * 4 + 3 maxes out at height * width * 4 - 1, which is size - 1.
+                    // Also, src_off maxes out at (height - 1) * stride, and so
+                    // src_off + x * 4 + 3 maxes out at (height - 1) * stride + width * 4 - 1.
+                    *buf.get_unchecked_mut(dst_off + x * 4 + 0) =
```

Comment:
if we're doing this I would like us to do one checked get at the end of our range before the loop, as a way of asserting that this is not UB.

Reply:
what I was thinking here is something like,

```rust
let data_len = height.saturating_sub(1) * stride + width * 4;
let data = { .. };
// all get_unchecked are in this range must succeed
assert!(data.get(data_len - 1).is_some());
```

In any case I don't think this is strictly necessary, but I find things like this helpful when reading the code? :shrug:

Reply:
Yup, that does look clearer. Thanks!